blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,149 @@
+	h := sha256.New()
+	h.Write(l[:])
+	h.Write(r[:])
+
+	buf := make([]byte, 0, 32)
+	first := h.Sum(buf)
+	h.Reset()
+	h.Write(first)
+	return *((*chainhash.Hash)(h.Sum(buf)))
+}
+
+// rollingMerkleTreeStore calculates the merkle root by only allocating O(logN)
+// memory where N is the total amount of leaves being included in the tree.
+type rollingMerkleTreeStore struct {
+	// roots are where the temporary merkle roots get stored while the
+	// merkle root is being calculated.
+	roots []chainhash.Hash
+
+	// numLeaves is the total leaves the store has processed.  numLeaves
+	// is required for the root calculation algorithm to work.
+	numLeaves uint64
+}
+
+// newRollingMerkleTreeStore returns a rollingMerkleTreeStore with the roots
+// allocated based on the passed in size.
+func newRollingMerkleTreeStore(size uint64) rollingMerkleTreeStore {
+	var alloc int
+	if size != 0 {
+		alloc = bits.Len64(size - 1)
+	}
+	return rollingMerkleTreeStore{roots: make([]chainhash.Hash, 0, alloc)}
+}
+
+// add adds a single hash to the merkle tree store.  Refer to algorithm 1 "AddOne" in
+// the utreexo paper (https://eprint.iacr.org/2019/611.pdf) for the exact algorithm.
+func (s *rollingMerkleTreeStore) add(add chainhash.Hash) {
+	// We can tell where the roots are by looking at the binary representation
+	// of the numLeaves.  Wherever there's a 1, there's a root.
+	//
+	// numLeaves of 8 will be '1000' in binary, so there will be one root at
+	// row 3. numLeaves of 3 will be '11' in binary, so there's two roots.  One at
+	// row 0 and one at row 1.
+	//
+	// In this loop below, we're looking for these roots by checking if there's
+	// a '1', starting from the LSB.  If there is a '1', we'll hash the root being
+	// added with that root until we hit a '0'.
+	newRoot := add
+	for h := uint8(0); (s.numLeaves>>h)&1 == 1; h++ {
+		// Grab the last root.
+		root := s.roots[len(s.roots)-1]
+
+		// Pop off the last root.
+		s.roots = s.roots[:len(s.roots)-1]
+
+		// Calculate the hash of the new root and append it.
+		newRoot = parentHash(root, newRoot)
+	}
+	s.roots = append(s.roots, newRoot)
+	s.numLeaves++
+}
+
+// calcMerkleRoot returns the merkle root for the passed in transactions.
+func (s *rollingMerkleTreeStore) calcMerkleRoot(adds []*btcutil.Tx, witness bool) chainhash.Hash {
+	for i := range adds {
+		// If we're computing a witness merkle root, instead of the
+		// regular txid, we use the modified wtxid which includes a
+		// transaction's witness data within the digest.  Additionally,
+		// the coinbase's wtxid is all zeroes.
+		switch {
+		case witness && i == 0:
+			var zeroHash chainhash.Hash
+			s.add(zeroHash)
+		case witness:
+			s.add(*adds[i].WitnessHash())
+		default:
+			s.add(*adds[i].Hash())
+		}
+	}
+
+	// If we only have one leaf, then the hash of that tx is the merkle root.
+	if s.numLeaves == 1 {
+		return s.roots[0]
+	}
+
+	// Add on the last tx again if there's an odd number of txs.
+	if len(adds) > 0 && len(adds)&1 == 1 {
```

Comment:
There doesn't seem to be anything conceptually different about these steps than the loop that follows below other than the fact that it is compensating for the APIs differing between Tx and Hash. I believe this function would be better served to be totally unaware of any Bitcoin mechanics and just operate on hashes. Instead make the transformation from []Tx to []Hash prior to extracting the merkle root from one of these stores.

Reply:
Again, I think the only reason this isn't a singular loop is because you are combining the concerns of Bitcoin specific digest calculations and the actual extraction of a full merkle root. The hash duplication operation is the same at all heights of the calculation, irrespective of whether or not the hash is a leaf hash or not. The fact that this function needs a leaf case doesn't make a lot of sense if you treat it as a full hash tree. At that point you only have your base case which is if there's only one root.

Reply:
> Again, I think the only reason this isn't a singular loop is because you are combining the concerns of Bitcoin specific digest calculations and the actual extraction of a full merkle root. The hash duplication operation is the same at all heights of the calculation, irrespective of whether or not the hash is a leaf hash or not. The fact that this function needs a leaf case doesn't make a lot of sense if you treat it as a full hash tree. At that point you only have your base case which is if there's only one root.

Could you write the code change you suggest here? Asking as simply getting rid of lines 95-103 won't work (as in that it won't calculate the correct roots) and am unsure what you mean.

Reply:
Yeah I'll PR into your branch the change that I think would simplify this.

Reply:
FYI this is nonblocking but I still have it on my list to help with this.

Reply:
Won't be addressed in this PR